server/resp/src/main/resources/resp.gr

```diff
@@ -0,0 +1,78 @@
+// RESP 3.x protocol
+namespace resp3x;
+
+// Target class
+class org.infinispan.server.resp.RespDecoder extends BaseRespDecoder;
+
+// static final field values are literals
+constants org.infinispan.server.resp.RespConstants;
+
+// methods simpleString, bulkString, byteArray, byte...
+intrinsics org.infinispan.server.resp.Intrinsics;
+
+import java.nio.charset.StandardCharsets;
+import java.lang.UnsupportedOperationException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletionStage;
+
+import io.netty.handler.codec.ByteToMessageDecoder;
+
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.commons.util.Util;
+import org.infinispan.server.resp.logging.Log;
+import org.infinispan.server.resp.Intrinsics.Resp2LongProcessor;
+import org.infinispan.util.concurrent.CompletionStages;
+
+init {
+   private final List<byte[]> reusedList = new ArrayList<>(16);
+
+   public RespDecoder(RespRequestHandler initialHandler) {
+      this.requestHandler = initialHandler;
+   }
+}
+
+beforedecode {
+   // We cannot read more than one command at a time
+   if (!ctx.channel().config().isAutoRead()) {
+      log.tracef("Auto read was disabled, not reading next bytes");
+      return;
+   } else {
+      log.tracef("Auto read was enabled, reading next bytes");
+   }
+}
+
+// this is the root
+root request
+   : singleByte { if (singleByte != ARRAY) throw new UnsupportedOperationException("Only array types are supported, received: " + singleByte); reusedList.clear(); }
+     number
+     singleByte
+     command
+     arguments #readNumber ( singleByte array { arguments.add(array); } )
+     execute
+   ;
+
+command returns String switch singleByte
+   : { BULK_STRING }? bulkString[longProcessor] { bulkString }
+   | { SIMPLE_STRING }? simpleString { simpleString }
+   | { throw new UnsupportedOperationException("Type not supported: " + singleByte); }
+   ;
+number: readNumber[longProcessor] { readNumber -= 1; };
+array returns byte[] switch singleByte
+   : { BULK_STRING }? bulkArray[longProcessor] { bulkArray }
+   // TODO: optimize this to not convert from bytes to string and back
```

Comment:
Running locally to test this with:

```java
public static byte[] simpleStringByte(ByteBuf buf) {
      // Find the end LF (would be nice to do this without iterating twice)
      int offset = buf.forEachByte(ByteProcessor.FIND_LF);
      if (offset <= 0 || buf.getByte(offset - 1) != '\r') {
         return null;
      }
      byte[] data = new byte[offset - 2 - buf.readerIndex()];
      buf.readBytes(data);
      buf.skipBytes(TERMINATOR_LENGTH);
      return data;
}
```

Reply:
Yeah, I was thinking something like this, but simple string isn't used very much so I didn't end up looking that closely. But this should work afaik.

Also another thing is I was looking at the simple string spec in the RESP protocol and it explicitly says you can't have a LF without the CR as well, so I think instead of returning null, we should probably throw a parsing exception.

> Simple Strings are encoded as follows: a plus character, followed by a string that cannot contain a CR or LF character (no newlines are allowed), and terminated by CRLF (that is "\r\n").

Reply:
Makes sense to me to return an error in such cases.